File: lib/utils/moon_calculator.dart
import 'package:apsl_sun_calc/apsl_sun_calc.dart';

/// Datos calculados del ciclo lunar para un día y ubicación.
class MoonData {
  final DateTime? moonrise;
  final DateTime? moonset;
  final double phase;        // 0.0-1.0 (0=nueva, 0.5=llena)
  final double illumination; // 0.0-1.0 fracción iluminada
  final String phaseName;

  MoonData({
    this.moonrise,
    this.moonset,
    required this.phase,
    required this.illumination,
    required this.phaseName,
  });

  @override
  String toString() => 'MoonData(phase: $phaseName, illumination: ${(illumination * 100).round()}%, '
      'rise: $moonrise, set: $moonset)';
}

/// Utilidad para calcular datos lunares usando apsl_sun_calc.
class MoonCalculator {
  /// Calcula los datos lunares para una fecha y ubicación.
  static MoonData calculate(DateTime date, double lat, double lng) {
    // Iluminación y fase
    final illum = SunCalc.getMoonIllumination(date);
    final phase = illum['phase'] as double;
    final fraction = illum['fraction'] as double;

    // Nombre de la fase
    final phaseName = _phaseName(phase);

    // Moonrise y moonset por búsqueda iterativa
    final moonrise = _findMoonEvent(date, lat, lng, true);
    final moonset = _findMoonEvent(date, lat, lng, false);

    return MoonData(
      moonrise: moonrise,
      moonset: moonset,
      phase: phase,
      illumination: fraction,
      phaseName: phaseName,
    );
  }

  /// Busca el momento en que la luna cruza el horizonte mediante búsqueda iterativa.
  /// [isRise] true para moonrise, false para moonset.
  static DateTime? _findMoonEvent(DateTime date, double lat, double lng, bool isRise) {
    // Buscamos en ventana de 24h desde medianoche local
    final startOfDay = DateTime(date.year, date.month, date.day);

    // Muestreo cada 10 minutos para encontrar el cruce
    const stepMinutes = 10;
    const totalSteps = (24 * 60) ~/ stepMinutes;

    double? prevAltitude;
    DateTime? crossingStart;

    for (int i = 0; i <= totalSteps; i++) {
      final t = startOfDay.add(Duration(minutes: i * stepMinutes));
      final pos = SunCalc.getMoonPosition(t, lat, lng);
      final altitude = pos['altitude'] as double;

      if (prevAltitude != null) {
        // Detectar cruce del horizonte (altitud = ~-0.01 rad para refracción)
        const horizon = -0.0145; // ~0.833° bajo horizonte por refracción atmosférica
        if (isRise && prevAltitude <= horizon && altitude > horizon) {
          crossingStart = t.subtract(Duration(minutes: stepMinutes));
          break;
        }
        if (!isRise && prevAltitude >= horizon && altitude < horizon) {
          crossingStart = t.subtract(Duration(minutes: stepMinutes));
          break;
        }
      }
      prevAltitude = altitude;
    }

    if (crossingStart == null) return null;

    // Refinar con búsqueda binaria (precisión ~1 minuto)
    var lo = crossingStart;
    var hi = crossingStart.add(Duration(minutes: stepMinutes));

    for (int i = 0; i < 8; i++) {
      final mid = lo.add(Duration(
        milliseconds: hi.difference(lo).inMilliseconds ~/ 2,
      ));
      final pos = SunCalc.getMoonPosition(mid, lat, lng);
      final alt = pos['altitude'] as double;
      const horizon = -0.0145;

      if (isRise) {
        if (alt <= horizon) {
          lo = mid;
        } else {
          hi = mid;
        }
      } else {
        if (alt >= horizon) {
          lo = mid;
        } else {
          hi = mid;
        }
      }
    }

    return lo.add(Duration(
      milliseconds: hi.difference(lo).inMilliseconds ~/ 2,
    ));
  }

  /// Nombre de la fase lunar en español.
  static String _phaseName(double phase) {
    if (phase < 0.03 || phase > 0.97) return 'Luna nueva';
    if (phase < 0.22) return 'Creciente cóncava';
    if (phase < 0.28) return 'Cuarto creciente';
    if (phase < 0.47) return 'Creciente convexa';
    if (phase < 0.53) return 'Luna llena';
    if (phase < 0.72) return 'Menguante convexa';
    if (phase < 0.78) return 'Cuarto menguante';
    return 'Menguante cóncava';
  }
}

File: app/src/main/java/com/nubo/nubo/domain/astro/MoonCalculator.kt
package com.nubo.nubo.domain.astro

import java.time.Duration
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId

/** Datos del ciclo lunar para un día y una ubicación. */
data class MoonData(
    val moonrise: LocalDateTime?,
    val moonset: LocalDateTime?,
    /** Posición en el ciclo: 0 = nueva, 0.5 = llena. */
    val cycle: Double,
    /** Fracción del disco iluminada, 0..1. */
    val illumination: Double,
    /** Fase con la que se nombra esa posición. */
    val phase: MoonPhase,
)

/** Calcula fase, iluminación y orto/ocaso lunar. */
object MoonCalculator {

    /** Altitud considerada horizonte, ya con la refracción atmosférica. */
    private const val HORIZON = -0.0145

    /** Paso del muestreo grueso al buscar el cruce del horizonte. */
    private const val STEP_MINUTES = 10L

    private const val DAY_HOURS = 24L

    /**
     * Ventana en la que buscar el ocaso tras el orto.
     *
     * La luna no aguanta sobre el horizonte más de unas trece horas fuera de
     * latitudes polares, así que con veintiséis sobra y no se cuela el orto
     * siguiente.
     */
    private const val ABOVE_HORIZON_HOURS = 26L

    fun calculate(
        dateTime: LocalDateTime,
        lat: Double,
        lng: Double,
        zone: ZoneId = ZoneId.systemDefault(),
    ): MoonData {
        val illumination = SunCalc.moonIllumination(dateTime.atZone(zone).toInstant())
        val date = dateTime.toLocalDate()

        val todayRise = findMoonEvent(date.atStartOfDay(), DAY_HOURS, lat, lng, true, zone)
        val yesterdayRise =
            findMoonEvent(date.minusDays(1).atStartOfDay(), DAY_HOURS, lat, lng, true, zone)

        // La luna está ahora mismo por encima del horizonte.
        val aloft = SunCalc.moonAltitude(dateTime.atZone(zone).toInstant(), lat, lng) > HORIZON

        val moonrise = when {
            // El trayecto que se dibuja es **el que está en curso**, y a
            // primera hora ese empezó ayer: la luna sale cada día unos
            // cincuenta minutos más tarde, así que de madrugada suele estar
            // arriba desde la víspera mientras el orto de hoy aún no ha
            // llegado. Cogiendo el de hoy, el arco describía un paso que no
            // había empezado y la tarjeta se quedaba sin punto de posición
            // teniendo la luna en el cielo.
            todayRise != null && dateTime < todayRise && aloft ->
                yesterdayRise ?: todayRise

            // Si la luna salió antes de medianoche no hay cruce ascendente hoy,
            // así que se mira el día anterior.
            else -> todayRise ?: yesterdayRise
        }

        // El ocaso se busca **a partir del orto**, no dentro del día natural.
        // Buscándolo desde medianoche se cogía el de la madrugada, que es el
        // final de la noche anterior: para una luna que sale a las 16:35 y se
        // pone a la 01:53 del día siguiente, devolvía un ocaso *anterior* al
        // orto, y con el intervalo invertido no había ni trayecto ni posición.
        val moonset = moonrise
            ?.let { findMoonEvent(it, ABOVE_HORIZON_HOURS, lat, lng, false, zone) }
            ?: findMoonEvent(date.atStartOfDay(), DAY_HOURS, lat, lng, false, zone)

        return MoonData(
            moonrise = moonrise,
            moonset = moonset,
            cycle = illumination.phase,
            illumination = illumination.fraction,
            phase = MoonPhases.phaseOn(date, zone, illumination.phase),
        )
    }

    /**
     * Busca cuándo cruza la luna el horizonte en las [hours] siguientes a
     * [from].
     *
     * Primero recorre el día a intervalos de 10 minutos hasta detectar el
     * cambio de signo, y después afina ese intervalo por bisección. No hay
     * fórmula cerrada para el orto lunar, así que este es el método habitual.
     */
    private fun findMoonEvent(
        from: LocalDateTime,
        hours: Long,
        lat: Double,
        lng: Double,
        rising: Boolean,
        zone: ZoneId,
    ): LocalDateTime? {
        val startOfDay = from
        val totalSteps = (hours * 60 / STEP_MINUTES).toInt()

        var previousAltitude: Double? = null
        var crossingStart: LocalDateTime? = null

        for (i in 0..totalSteps) {
            val t = startOfDay.plusMinutes(i * STEP_MINUTES)
            val altitude = SunCalc.moonAltitude(t.atZone(zone).toInstant(), lat, lng)

            val previous = previousAltitude
            if (previous != null) {
                val crossed = if (rising) {
                    previous <= HORIZON && altitude > HORIZON
                } else {
                    previous >= HORIZON && altitude < HORIZON
                }
                if (crossed) {
                    crossingStart = t.minusMinutes(STEP_MINUTES)
                    break
                }
            }
            previousAltitude = altitude
        }

        val start = crossingStart ?: return null

        // Bisección: 8 iteraciones sobre un intervalo de 10 min dejan el
        // resultado por debajo del segundo, de sobra para mostrar una hora.
        var low = start
        var high = start.plusMinutes(STEP_MINUTES)
        repeat(8) {
            val mid = low.plus(Duration.between(low, high).dividedBy(2))
            val altitude = SunCalc.moonAltitude(mid.atZone(zone).toInstant(), lat, lng)
            val keepLow = if (rising) altitude <= HORIZON else altitude >= HORIZON
            if (keepLow) low = mid else high = mid
        }

        return low.plus(Duration.between(low, high).dividedBy(2))
    }

}
